torchrl/gradient_collector.py

```diff
@@ -0,0 +1,168 @@
+                if optimizer.state.get(name) is None:
+                    warnings.warn(f"{name} is NOT being tracked by the optimizer")
+
+    def __iter__(self) -> Iterator[TensorDictBase]:
+        return self.iterator()
+
+    def next(self):
+        try:
+            if self._iterator is None:
+                self._iterator = iter(self)
+            out = next(self._iterator)
+            # if any, we don't want the device ref to be passed in distributed settings
+            out.clear_device_()
+            return out
+        except StopIteration:
+            return None
+
+    def update_policy_weights_(
+            self,
+            policy_weights: Optional[TensorDictBase] = None,
+            critic_weights: Optional[TensorDictBase] = None,
+            objective_weights: Optional[TensorDictBase] = None,
+    ) -> None:
+
+        # TODO: is this the right way to do it ?
+        # Update policy and critic
+        self.objective_params.apply(lambda x: x.data).update_(objective_weights)
+        self.policy_params.apply(lambda x: x.data).update_(policy_weights)
+        self.critic_params.apply(lambda x: x.data).update_(critic_weights)
+
+        for name, param in self.objective.named_parameters():
+            param.data = objective_weights.get(name)
+        for name, param in self.policy.named_parameters():
+            param.data = policy_weights.get(name)
+        for name, param in self.critic.named_parameters():
+            param.data = critic_weights.get(name)
+
+        self.collector.update_policy_weights_()
+
+    def shutdown(self):
+        self.collector.shutdown()
+
+    def iterator(self) -> Iterator[TensorDictBase]:
+        grads = self._step_iterator()
+        return grads
+
+    def _step_iterator(self):
+        """Computes next gradient in each iteration."""
+
+        for data in self.collector:
+
+            # Add to replay buffer
+            self.replay_buffer.extend(data)
+
+            for _ in range(self.updates_per_batch):
+
+                # Sample batch from replay buffer
+                mini_batch = self.replay_buffer.sample().to(self.device)
+
+                # Compute loss
+                loss = self.objective(mini_batch)
+                loss_sum = sum([item for key, item in loss.items() if key.startswith("loss")])
+
+                # Backprop loss
+                self.optimizer.zero_grad()
+                loss_sum.backward()
+
+                # Get gradients as a Tensordict
+                objective_params = TensorDict(dict(self.objective.named_parameters()), [])
+                objective_grads = objective_params.apply(lambda x: x.grad)
+                actor_params = TensorDict(dict(self.policy.named_parameters()), [])
+                actor_grads = actor_params.apply(lambda x: x.grad)
+                critic_params = TensorDict(dict(self.critic.named_parameters()), [])
+                critic_grads = critic_params.apply(lambda x: x.grad)
```

Comment:
```suggestion
                objective_params = TensorDict(dict(self.objective.named_parameters()), [])
                objective_grads = objective_params.apply(lambda x: x.grad)
```

Only the first two lines should be necessary

Not sure how that would work if the objective_params also contain buffers. Maybe we should do

```
objective_grads = objective_params.apply(lambda x: x.grad if x.grad is not None else torch.zeros_like(x))
```

but it could be unsafe (e.g. if by mistake a param is ignored)